addons/website_portal_sale/models/sale_order.py

```diff
@@ -47,6 +59,34 @@ def _notification_recipients(self, message, groups):
 
         return groups
 
+    @api.model_cr_context
+    def _init_column(self, column_name):
+        """ Initialize the value of the given column for existing rows.
+
+            Overridden here because we skip generating unique access tokens
+            for potentially tons of existing sales orders that won't need it.
+            We do, however, generate it for draft or quotation sales orders.
+        """
+        if column_name != 'access_token':
+            super(SaleOrder, self)._init_column(column_name)
+        else:
+            default_compute = self._fields[column_name].default
+
+            query = """
+                SELECT id FROM "%s"
+                 WHERE "%s" is NULL
+                       AND state in %%s
+            """ % (self._table, column_name)
+            self.env.cr.execute(query, (('draft', 'sent'),))
+            order_ids = self.env.cr.fetchall()
+
+            for order_id, in order_ids: # order_ids is a list of tuples e.g.: [(1,), (2,), ...]
+                default_value = default_compute(self)
+
+                query = 'UPDATE "%s" SET "%s"=%%s WHERE id = %s' % (
+                    self._table, column_name, order_id)
```

Comment:
Sorry, forgot to reply to this part. Making identifiers parameters is indeed not necessary, except for abstract classes like `mail.thread` where the code might need to refer to different tables. Elsewhere it makes the code noisy and error-prone. People tend to confuse the different kinds of parameters (identifiers who need careful injection vs value parameters that must be passed as query parameters to the execute method), and c/p mistakes. That was the case here. So let's keep things as simple as possible.
And have people [learn the API inside out](http://initd.org/psycopg/docs/usage.html#the-problem-with-the-query-parameters) asasp ;-)

PS: @Icallhimtest out of curiosity, where did you originally copy this from? That place might need to be fixed as well ;-)